tests/Collection/CodecCollectionFunctionalTest.php

```diff
@@ -0,0 +1,528 @@
+        ];
+
+        yield 'No codec' => [
+            'expected' => self::createObjectFixtureResult(1, true),
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => Document::fromPHP(self::createObjectFixtureResult(1, true)),
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideInsertOneOptions */
+    public function testInsertOne($expected, $options): void
+    {
+        $result = $this->collection->insertOne(TestObject::createForFixture(1), $options);
+        $this->assertSame(1, $result->getInsertedCount());
+
+        if ($expected instanceof BSONDocument && $expected->_id === null) {
+            $expected->_id = $result->getInsertedId();
+        }
+
+        if ($expected instanceof Document && $expected->get('_id') === null) {
+            $inserted = $expected->toPHP();
+            $inserted->_id = $result->getInsertedId();
+            $expected = Document::fromPHP($inserted);
+        }
+
+        $this->assertEquals($expected, $this->collection->findOne([], $options));
+    }
+
+    public static function provideReplaceOneOptions(): Generator
+    {
+        $replacedObject = TestObject::createForFixture(1, true);
+        $replacedObject->x->foo = 'baz';
+
+        yield 'Default codec' => [
+            'expected' => $replacedObject,
+            'options' => [],
+        ];
+
+        $replacedObject = self::createObjectFixtureResult(1);
+        $replacedObject->x->foo = 'baz';
+
+        yield 'No codec' => [
+            'expected' => $replacedObject,
+            'options' => ['codec' => null],
+        ];
+
+        yield 'BSON type map' => [
+            'expected' => Document::fromPHP($replacedObject),
+            'options' => ['typeMap' => ['root' => 'bson']],
+        ];
+    }
+
+    /** @dataProvider provideReplaceOneOptions */
+    public function testReplaceOne($expected, $options): void
+    {
+        $this->createFixtures(1);
+
+        $replaceObject = TestObject::createForFixture(1);
+        $replaceObject->x->foo = 'baz';
+
+        $result = $this->collection->replaceOne(['_id' => 1], $replaceObject, $options);
+        $this->assertSame(1, $result->getMatchedCount());
+        $this->assertSame(1, $result->getModifiedCount());
+
+        $this->assertEquals($expected, $this->collection->findOne([], $options));
+    }
+
+    public function testReplaceOneWithCodecAndTypemap(): void
+    {
+        $options = [
+            'codec' => new TestDocumentCodec(),
+            'typeMap' => ['root' => 'array', 'document' => 'array'],
+        ];
+
+        $this->expectExceptionObject(InvalidArgumentException::cannotCombineCodecAndTypeMap());
+        $this->collection->replaceOne(['_id' => 1], ['foo' => 'bar'], $options);
+    }
+
+    /**
+     * Create data fixtures.
+     */
+    private function createFixtures(int $n, array $executeBulkWriteOptions = []): void
```

Comment:
Noted that this is another place where you coupled the structure of TestObject into a fixture creation method. I'd suggest making TestObject directly Serializable to work around this, and you can leave a comment on the `bsonSerialize()` method explaining why you do so.

----

One thing I didn't consider is whether you intentionally _didn't_ implement Serializable to ensure that the document is only encoded correctly when a codec is provided to the operation.

If that's the case then I think we can work around this by introducing a new fixture creation method in TestObject that returns a Document instance (using `Document::fromPHP()`) that we can pass directly to `BulkWrite::insert()`. That would still allow us to consolidate the TestObject schema to its own class without scattering it throughout a few functional tests' `createFixtures()` methods.

Reply:
Not making it `Serializable` was on purpose, as the latter would make errors less explicit when inserting or updating such documents without a codec. I've added a factory to create a `Document` instance for use in the `createFixtures` methods.